### DISCUSSION TITLE
random core dump caused by ilegal memory free

### DISCUSSION THREAD

**Discussion by kevisminename** [libvips/libvips#4074] [Help wanted]
Greetings, libvips is such a great image library that we used it to produce lots of images with text in it, but recently we noticed that core(SIGSEGV) has been happening for a long time since we used it (about half of year ago, it happened in a quite odd rhythm, and it did't affect our production, hence we found out this problem this late), and out of curiousty and future safety concern, we decided to take a look at it, here is the info we found out:

- **system envirinment:** 

ubuntu  20.04.1 LTS (Focal Fossa) in a docker container

- **vips version:** 

libvips-8.14.3, relase version (we did't compile it by ourself, instead downloaded it from open resource that can not be traced anymore)

- **runtime system infomation when cores:**

cpu at average 20%-40% with 8Cs
memory at average around 40- 60% with 8G, later we upgraded the memory to 32G, memory stays around 10%-20%.

- **core frequency:** 

ussually one time per approxmitely one million images processed, put it in time, will happen once every 20-30hours at 10 qps, but some time occassionly, it cores after several days.

- **is reproducible?** 

not reproducible, every time we can locate the exact input that caused the core, but under new environment, it can be processed smoothly without any problem. we even collected all the inputs that was being procesed before, during, and after the core, still did't reproduce.

- **core trace**

![begin](https://github.com/user-attachments/assets/49d53bc7-8a4c-42c5-83ac-b418a99dc394)
![middle](https://github.com/user-attachments/assets/2fa43e4b-fbbd-4d3f-a971-7de0cf14ae29)
![end](https://github.com/user-attachments/assets/664a118a-a2ad-47f7-9b9c-cd7c5e6e4023)

All the core traces we met boils down with the same end, libc's malloc free (free invalid pointer).

**- our code example:**

```C++
//internal data object that has its own render material and common render functions
std::vector<* INTERNAL_OBJ> reqs;

try
{
    std::vector<vips::VImage> tmpImages;
    for (int i = 0; i < reqs.size(); i++)
    {
        tmpImages.push_back((reqs.at(i))->process());
    }

    vips::VImage resultImage = tmpImages[0].copy();
    for (int i = 1; i < reqs.size(); i++)
    {
        resultImage = resultImage.join(tmpImages[i], VipsDirection::VIPS_DIRECTION_VERTICAL, vips::VImage::option()->set("expand", true)->set("align", VipsAlign::VIPS_ALIGN_LOW));
    }

    void *buffer = nullptr;
    size_t len = 0;
    if (!reqFormat.compare("jpg") || !reqFormat.compare("JPG"))
    {
        resultImage.write_to_buffer(".jpg", &buffer, &len, vips::VImage::option()->set("Q", 95));
    }
    else
    {
        resultImage.write_to_buffer(".png", &buffer, &len, vips::VImage::option()->set("compression", 1)->set("filter", VIPS_FOREIGN_PNG_FILTER_ALL));
    }

    while (!reqs.empty())
    {
        delete reqs.back();
        reqs.pop_back();
    }

}
catch (const vips::VError &err)
{
    LOG(ERROR) << err.what();
    LOG(ERROR) << "vips process err request id is : " << requestId << std::endl;
}
```

this try block is in a warper of rpc server, it gets to be called in paralle, so there could be more than a dozen of threads running this part of code.
core happened when the try block finisnes and the `tmpImages` auto-destroys itselves,  the decode part should prove that tmpImages' content is alright.

- **what we tried**

We tried to compile libvips-8.14.3 in debug mode, and managed to get the debug info library in the production, but the image process became extremely slow, that we can not reach 10 qps, and no core was spotted.
so there are no detailed debug trace of the destruct function of VImage.

Without the knowledge of what Vimage's being processed and destroyed, our current guess it that it may be a memory double free problem caused by data race that only happened with certain parallel condition. 

So we were hoping that anyone who might encount this problem before or who might have the sufficient knowledge of Vimage would throw us a pice of information or instructions that what might happen here and what should we do next to find our more infomation.
anything would be much appreciated.

btw
we did find and try out issue 401(https://github.com/lovell/sharp/issues/401) 's method, but it did't help out.

**Comment by jcupitt** [comment]
Hello @kevisminename,

That looks like the stack trace you get with a double unref -- perhaps some input buffer is being unreffed before the output image has completely finished writing?

Could you work this code sample up into something I could run? I can't see how the input images are being made from this code sample, for example.

> `vips::VImage resultImage = tmpImages[0].copy();`

You don't need the `copy()` here, though I don't think that would make any difference.

**Reply by kevisminename** [reply]
Thx for your time!
I tried to find out what happened with VImage::Copy(), but it was too complicated, only by comments of other funcitons, I made a  guess that C++ function obj.copy() is a shallow copy(data buffer is not actually copied), and reference count was used.
why do you think it would make no difference, could it be that resultImage and tmpImages' auto-destruct conflicted and caused the core?  we will try to optmize the resultImage part of code and give it a try.

**Reply by jcupitt** [reply]
`VImage` is a smart pointer, so you can just do:

```C++
VImage a = b;
```

and it'll adjust all the ref counts for you automatically. There's no need for an explicit `.copy()` operation.

Adding an extra `.copy()` is harmless, it's just that you don't have to do that.

**Reply by kevisminename** [reply]
I have a question, if its a duble ref problem, then is it the vips' reference count scheme that is not working, or is it our client's problem (let's sat we misused of vips function or illegal operated of VImage object which messes up the reference count)?

**Reply by jcupitt** [reply]
It's probably a bug in your code, but of course I can't be certain.

**Comment by jcupitt** [comment]
... and I would not use a libvips binary I found from somewhere I can't trace :) 

I would use the platform libvips. 20.04 ships with 8.9.1, which (I think?) ought to work OK for this code. If you use the official binary you can be reasonably sure there isn't some incompatibility with your system libraries. This could possibly be the cause of your crash.

You can also build current stable libvips from source for 20.04. This is pretty easy in docker, for example:

https://github.com/jcupitt/docker-builds/tree/master/libvips-ubuntu20.04

**Reply by kevisminename** [reply]
yeah, your point is  well understood, we wil definitely try out your recommended version of librarys, so to clean the suspect of our lib.

**Reply by kevisminename** [reply]
@jcupitt  sir, I tried your docker and it compiled  successfully, but the so did't work because error of "VipsOperation: class "text" not found", I checked issue https://github.com/libvips/libvips/issues/336, it said something wrong with pango, so it seemed your docker is not the "complete dev" version, is this right? where can i find a docker environment that all the relys of vips is available?

**Reply by jcupitt** [reply]
Yes, I only included a selection of the possible dependencies in my dockerfile. You'll have to adjust them if you need things like text rendering.

**Reply by jcupitt** [reply]
I added pango to the dockerfile, removed the libdir option, and added an ldconfig. I now see:

```
$ docker build -t libvips-ubuntu20.04 .
... stuff
$ docker run -it --rm libvips-ubuntu20.04                                            
root@f253756942ce:/usr/local/src# vips text
make a text image
usage:
   text out text [--option-name option-value ...]
where:
   out          - Output image, output VipsImage
   text         - Text to render, input gchararray
...
```
